google/cloud/spanner/internal/connection_impl_test.cc

```diff
@@ -801,23 +867,84 @@ TEST(ConnectionImplTest, QueryOptions) {
+    {
+      InSequence seq;
+
+      auto execute_sql_request_matcher =
+          AllOf(Property(&spanner_proto::ExecuteSqlRequest::query_options,
+                         IsProtoEqual(tc.qo_proto)),
+                Property(&spanner_proto::ExecuteSqlRequest::request_options,
+                         IsProtoEqual(tc.ro_proto)));
+      auto begin_transaction_request_matcher =
+          Property(&spanner_proto::BeginTransactionRequest::request_options,
+                   IsProtoEqual(spanner_proto::RequestOptions{}));
+      auto execute_batch_dml_request_matcher =
+          Property(&spanner_proto::ExecuteBatchDmlRequest::request_options,
+                   IsProtoEqual(tc.ro_proto));
+      auto read_request_matcher =
+          Property(&spanner_proto::ReadRequest::request_options,
+                   IsProtoEqual(tc.ro_proto));
+      auto commit_request_matcher =
+          Property(&spanner_proto::CommitRequest::request_options,
+                   IsProtoEqual(tc.ro_proto));
+
+      // ExecuteQuery().
+      EXPECT_CALL(*mock, BatchCreateSessions(_, HasDatabase(db)))
+          .WillOnce(Return(MakeSessionsResponse({"session-name"})))
+          .RetiresOnSaturation();
+      EXPECT_CALL(*mock, ExecuteStreamingSql(_, execute_sql_request_matcher))
+          .WillOnce(Return(ByMove(std::move(grpc_reader))))
+          .RetiresOnSaturation();
+
+      // ProfileQuery().
+      EXPECT_CALL(*mock, ExecuteStreamingSql(_, execute_sql_request_matcher))
+          .WillOnce(
+              Return(ByMove(absl::make_unique<NiceMock<MockGrpcReader>>())))
+          .RetiresOnSaturation();
+
+      // ExecutePartitionedDml().
+      EXPECT_CALL(*mock, BatchCreateSessions(_, HasDatabase(db)))
+          .WillOnce(Return(MakeSessionsResponse({"session-name"})))
+          .RetiresOnSaturation();
+      EXPECT_CALL(*mock, BeginTransaction(_, begin_transaction_request_matcher))
+          .WillOnce(Return(MakeTestTransaction()))
+          .RetiresOnSaturation();
+      EXPECT_CALL(*mock, ExecuteStreamingSql(_, execute_sql_request_matcher))
+          .WillOnce(
+              Return(ByMove(absl::make_unique<NiceMock<MockGrpcReader>>())))
+          .RetiresOnSaturation();
+
+      // ExecuteDml(), ProfileDml(), AnalyzeSql().
+      EXPECT_CALL(*mock, ExecuteSql(_, execute_sql_request_matcher))
+          .Times(3)
+          .RetiresOnSaturation();
+
+      // ExecuteBatchDml().
+      EXPECT_CALL(*mock, ExecuteBatchDml(_, execute_batch_dml_request_matcher))
+          .Times(1)
+          .RetiresOnSaturation();
+
+      // Read().
+      EXPECT_CALL(*mock, StreamingRead(_, read_request_matcher))
+          .WillOnce(
+              Return(ByMove(absl::make_unique<NiceMock<MockGrpcReader>>())))
+          .RetiresOnSaturation();
+
+      // Commit().
+      EXPECT_CALL(*mock, Commit(_, commit_request_matcher))
+          .Times(1)
+          .RetiresOnSaturation();
+    }
+
+    (void)conn->ExecuteQuery(sql_params);
```

Comment:
Maybe these should be separate tests, but I can see how that would be a lot more boilerplate code. No action needed.